File: solutions/haskell/perfect-numbers/1/src/PerfectNumbers.hs
module PerfectNumbers (classify, Classification(..)) where

data Classification = Deficient | Perfect | Abundant deriving (Eq, Show)

divisors :: Int -> [Int]
divisors n = [x | x <- [1 .. (n -1)], n `rem` x == 0]

aliquot :: Int -> Int
aliquot = sum . divisors

classify :: Int -> Maybe Classification
classify n
  | n <= 0 = Nothing
  | a < n = Just Deficient
  | a > n = Just Abundant
  | a == n = Just Perfect
  | otherwise = Nothing
  where
    a = aliquot n

File: solutions/elixir/perfect-numbers/1/lib/perfect_numbers.ex
defmodule PerfectNumbers do
  @doc """
  Determine the aliquot sum of the given `number`, by summing all the factors
  of `number`, aside from `number` itself.

  Based on this sum, classify the number as:

  :perfect if the aliquot sum is equal to `number`
  :abundant if the aliquot sum is greater than `number`
  :deficient if the aliquot sum is less than `number`
  """
  @spec classify(number :: integer) :: {:ok, atom} | {:error, String.t()}
  def classify(number) when number <= 0,
    do: {:error, "Classification is only possible for natural numbers."}

  def classify(number) do
    aliquot_sum = factors(number)

    cond do
      aliquot_sum < number -> {:ok, :deficient}
      aliquot_sum > number -> {:ok, :abundant}
      true -> {:ok, :perfect}
    end
  end

  defp factors(n),
    do:
      1..(n |> :math.sqrt() |> trunc())
      |> Enum.reduce([], &(&2 ++ if(rem(n, &1) == 0, do: [&1, div(n, &1)], else: [])))
      |> Enum.uniq()
      |> List.delete(n)
      |> Enum.sum()
end
